File: 1559-detect-cycles-in-2d-grid/1559-detect-cycles-in-2d-grid.py
class Solution:
    def __init__(self):
        self.dir = [(1, 0), (-1, 0), (0, 1), (0, -1)]
    def containsCycle(self, grid: List[List[str]]) -> bool:
        row=len(grid)
        col=len(grid[0])
        v=[[False]*col  for i in range(row)]
        def solve(i,j,prev_i,prev_j):
            v[i][j]=True
            for dx,dy in self.dir:
                nx=i+dx
                ny=j+dy
                if(nx<0 or ny<0 or nx>=row or ny>=col):
                    continue
                if (grid[i][j]!=grid[nx][ny]) :
                    continue
                if nx==prev_i and ny==prev_j:
                    continue
                if v[nx][ny]:
                    return True
                if solve(nx,ny,i,j):
                    return True

            return False
        
        for i in range(row):
            for j in range(col):
                if not v[i][j]:
                    if solve(i,j,-1,-1):
                        return True
        return False

File: 1559-detect-cycles-in-2d-grid/1559-detect-cycles-in-2d-grid.java
class Solution {
    int dir[][] = { { 1, 0 }, { -1, 0 }, { 0, -1 }, { 0, 1 } };

    public boolean containsCycle(char[][] grid) {
        int vis[][]=new int[grid.length][grid[0].length];
        for(int i=0;i<grid.length;i++){
            for(int j=0;j<grid[0].length;j++){
                if(vis[i][j]!=1)
{                if(solve(grid,-1,-1,i,j,vis)){
                    return true;
                }}
            }
        }
        return false;
        
    }

    boolean solve(char[][] grid,int prev_r,int prev_c,int r,int c,int v[][] ){
        v[r][c]=1;
        for (int d[]:dir){
            int nr=r+d[0];
            int ny=c+d[1];
            if(nr<0 || ny<0 || nr>=grid.length||ny>=grid[0].length){
                continue;
            }
            if((nr==prev_r&&ny==prev_c)||grid[r][c]!=grid[nr][ny]){
                continue;
            }
            if(v[nr][ny]==1){
                return true;
            }
            if(solve(grid,r,c,nr,ny,v)){
                    return true;
                }
        }
        return false;
    }
}